c++/src/kj/compat/tls.c++

```diff
@@ -405,31 +416,108 @@ private:
-      });
+    if (auto localFulfiller = kj::mv(fulfiller)) {
+      // Fast path: We have a consumer waiting.
+      localFulfiller->fulfill(kj::fwd<T>(v));
+      return;
+    }
+
+    // Slow path: We don't have a consumer, add it to the queue.
+    auto node = kj::heap<Node>(kj::fwd<T>(v));
+    if (head) {
+      Node* currentNode = head.get();
+      while (currentNode->next) {
+        // Run through the nodes until we don't have a next.
+        currentNode = currentNode->next.get();
+      }
+
+      currentNode->next = kj::mv(node);
+    } else {
+      head = kj::mv(node);
+    }
+  }
+
+  Promise<void> push(Promise<T>&& p) {
+    // Eventually push a value onto the queue.
+
+    return kj::fwd<Promise<T>>(p).then([this](T&& v) {
+      pushRaw(kj::fwd<T>(v));
     });
   }
 
-  Promise<kj::AuthenticatedStream> acceptAuthenticated() override {
-    return inner->acceptAuthenticated().then([this](kj::AuthenticatedStream stream) {
-      return kj::evalNow([&] { return tls.wrapServer(kj::mv(stream)); })
-          .catch_([this](kj::Exception&& e) {
-        KJ_LOG(ERROR, "error accepting tls connection", e);
-        return acceptAuthenticated();
-      });
+  Promise<T> pop() {
+    // Eventually pop a value from the queue.
+    // Note that if your sink lags your sources, the promise will always be ready.
+
+    KJ_REQUIRE(!fulfiller, "InputQueue can only have one waiter");
+
+    if (head) {
+      // Fast path: There are already values in queue for us to take.
+      auto value = kj::mv(head->value);
+      ;
+      head = kj::mv(head->next);
+      return kj::mv(value);
+    } else {
+      // Slow path: We need to wait for a value to arrive.
+      auto paf = newPromiseAndFulfiller<T>();
+      fulfiller = kj::mv(paf.fulfiller);
+
+      return kj::mv(paf.promise);
+    }
+  }
+
+private:
+  struct Node {
+    // A simple intrusive linked list node which extends the lifetime of the next node.
+
+    Node(Promise<T>&& value): next{}, value{ kj::fwd<Promise<T>>(value) } {
+      // Always construct with no next node to start.
+    }
+
+    Own<Node> next;
+    Promise<T> value;
+  };
+
+  Own<Node> head;
+  Own<PromiseFulfiller<T>> fulfiller;
+};
+
+class TlsConnectionReceiver final: public ConnectionReceiver, public TaskSet::ErrorHandler {
+public:
+  TlsConnectionReceiver(TlsContext& tls, Own<ConnectionReceiver> inner)
+      : tls(tls), inner(kj::mv(inner)), tasks(*this) {}
+
+  void taskFailed(Exception&& e) override {
+    KJ_LOG(ERROR, "error accepting tls connection", kj::fwd<Exception>(e));
+  };
+
+  Promise<Own<AsyncIoStream>> accept() override {
+    return acceptAuthenticated().then([](AuthenticatedStream&& stream) {
+      return kj::fwd<Own<AsyncIoStream>>(stream.stream);
     });
   }
 
+  Promise<AuthenticatedStream> acceptAuthenticated() override {
+    return queue.pop();
+  }
+
+  Promise<void> run() override {
+    // TODO(pr) We could get away without this function by adding the task in the ctor.
```

Comment:
I must admit that I lean towards defensive programming if it seems "free". I'm generally conflicted about this choice. On one end, it lays a foundation for a flexible accept loop in other ConnectionReceivers and I've see TODOs mentioning how that would be nice. It also allows for deferred init, which can be very nice for strictly ordered initialization. On the other hand, we could get away without changes in repos that use KJ if we start it in the ctor.

Reply:
Hmmmmm, so I can see two potentially nice options then:
- We can make accept()/acceptAuthenticated() turn do one accept loop if we haven't invoked run().
- We can provide a different method instead of run(). `Promise<void> onFatalError()` or something less scary sounding.